### PR TITLE
Serialize nested includes

### PR DESCRIPTION
**Pull request by jshow**
- Dish off to associated serializers when required

includes syntax is:
```
      options[:include] = [:actors, :'actors.agency']
      serializable_hash = MovieSerializer.new([movie], options).serializable_hash
```

### CODE REVIEW

File: lib/fast_jsonapi/serialization_core.rb [hunk 1]
```diff
@@ -99,21 +99,42 @@ def to_json(payload)
...
+          remaining_items = nil
+          if items.size > 1
+            items_copy = items.dup
+            items_copy.delete_at(0)
+            remaining_items = [items_copy.join('.').to_sym]
+          end
+          item_to_serialize = items.last
+
+          items.each do |item|
+            next unless relationships_to_serialize[item]
```
Comment (yalesov) [inline lib/fast_jsonapi/serialization_core.rb]: https://github.com/LifeTales/fast_jsonapi/blob/49e7e86e93c61fa5cf34278144378c8640d9f079/lib/fast_jsonapi/object_serializer.rb#L154-L155

`relationships_to_serialize` might be `nil`, throwing `undefined method `[]' for nil:NilClass`; need to catch that too.

Comment (yalesov) [reply lib/fast_jsonapi/serialization_core.rb]: made corresponding PR for this at https://github.com/LifeTales/fast_jsonapi/pull/1

**Comment by gschorkopf** [comment]
Can I help in rebasing anything to resolve 👇 merge conflict?

**Comment by jshow** [comment]
much appreciated @gschorkopf

**Comment by gschorkopf** [comment]
Just a heads up: I rebased locally and was ready to push, but got several errors in the speed test section of the suite. Along the lines of:

```
FastJsonapi::ObjectSerializer when comparing with AMS 0.10.x should serialize 1 records atleast 25 times faster than AMS
expected: >= 25
got:    22.57446706623082
```

If any of y'all have time today to rebase and investigate, that'd be swell. I can revisit tomorrow too.

**Comment by jshow** [comment]
thanks @gschorkopf - I get those performance related failures pulling master

**Comment by llenodo** [comment]
Anything blocking this PR? Would love to use this added functionality ASAP

**Comment by thommahoney** [comment]
@jshow could you resolve the conflicts on this branch at least? Unsure which of the maintainers to ping to get this out but the merge conflicts will block that progress.

**Comment by jshow** [comment]
@thommahoney , rebased against dev, we're gold now

**Comment by jshow** [comment]
@nathanstitt , wdyt ?

**Comment by nathanstitt** [comment]
@jshow looks good to me - I've been tracking the branch in [a new app I'm spiking](https://github.com/a15k/mothership) and it seems to be working well.

**Comment by gabrielmm1234** [comment]
When is it going to be merged?

**Comment by mhluska** [comment]
Does this currently work with polymorphic associations? `include: [:'order_items.item']` throws `NameError (uninitialized constant ItemSerializer)` for me. Not sure if related.

File: lib/fast_jsonapi/object_serializer.rb [hunk 2]
```diff
@@ -235,6 +237,25 @@ def fetch_polymorphic_option(options)
         return option if option.respond_to? :keys
         {}
       end
+
+      def get_serializer(item)
+        unless item.to_s.include?('.')
```
Comment (shishirmk) [inline lib/fast_jsonapi/object_serializer.rb]: The logic seems very similar for both parts of this method. Is there anyway we can combine them and make it more concise?

File: lib/fast_jsonapi/serialization_core.rb [hunk 3]
```diff
@@ -106,21 +106,42 @@ def to_json(payload)
...
-          next if included_objects.blank?
-          included_objects = [included_objects] unless relationship_type == :has_many
-          included_objects.each do |inc_obj|
-            code = "#{record_type}_#{inc_obj.id}"
-            next if known_included_objects.key?(code)
-            known_included_objects[code] = inc_obj
-            included_records << serializer.record_hash(inc_obj)
+        return unless includes_list.present?
+
+        includes_list.sort.each_with_object([]) do |include_item, included_records|
```
Comment (shishirmk) [inline lib/fast_jsonapi/serialization_core.rb]: I think we should split this method up into smaller methods. That would make it more readable and probably easier to write to unit tests for.

**Comment by jshow** [comment]
@shishirmk , I've simplified the validates of includes options, plus dry'ed up the parsing of nested includes

is this providing the kind of clarity you were looking for ? If not, please provide more specific direction

I'll work on the README once we've settled the code

This is a long running branch, getting tough to deal with merge conflicts, I'd appreciate a quick cycle so we can get this into dev. thanks

**Comment by shishirmk** [comment]
@jshow will take a look today and try to make this a quicker development cycle.

File: lib/fast_jsonapi/object_serializer.rb [hunk 4]
```diff
@@ -234,6 +224,20 @@ def fetch_polymorphic_option(options)
         return option if option.respond_to? :keys
         {}
       end
+
+      def validate_includes!(includes)
```
Comment (shishirmk) [inline lib/fast_jsonapi/object_serializer.rb]: This method looks good.

File: spec/lib/object_serializer_spec.rb [hunk 5]
```diff
@@ -30,6 +30,30 @@
       expect(serializable_hash[:included]).to be nil
     end
 
+    it 'returns correct nested includes when serializable_hash is called' do
```
Comment (shishirmk) [inline spec/lib/object_serializer_spec.rb]: We should add more tests

- test a has many
- test a has_one
- test a multi level case for example 'actors.agency.adresses'
- we should test this with a polymorphic relationship
- we should probably also test it with a relationship block https://github.com/Netflix/fast_jsonapi/pull/195

**Comment by shishirmk** [comment]
@jshow the code looks good. Just need more tests and as you mentioned we need to add some documentation

**Comment by jshow** [comment]
@shishirmk , added some tests and found a bug with `has_one` - it was treated like a `belongs_to` - and tests to support that bad assumption - probably should be a separate PR, but couldn't move forward here without it being fixed, so it's all here

in addition
* `test a has many` - done
* `test a has_one` - done (with the provisor above)
* `test a multi level case for example 'actors.agency.adresses'` - done
* `we should test this with a polymorphic relationship` - this test is failing, what is the expected result, there aren't any tests that I can based the expected behaviour on - https://github.com/Netflix/fast_jsonapi/pull/152/files#diff-70daca4f50c5b49e4a8e037801a4c455R202
* `we should probably also test it with a relationship block #195` - not attempted yet

This has turned in a monstrous effort - specific comments to help bring this to the finish line are appreciated

**Comment by jshow** [comment]
Thanks for your support @shishirmk. It's all yours!

Once this makes it to dev, I'll get the associated https://github.com/Netflix/fast_jsonapi/pull/161 ready

File: lib/fast_jsonapi/serialization_core.rb [hunk 6]
```diff
@@ -146,7 +172,11 @@ def fetch_id(record, relationship, params)
           return object.id
         end
 
-        record.public_send(relationship[:id_method_name])
+        if relationship[:relationship_type] == :has_one
+          record.public_send(relationship[:object_method_name])&.id
```
Comment (shishirmk) [inline lib/fast_jsonapi/serialization_core.rb]: This is causing a syntax error.

**Comment by jshow** [comment]
@shishirmk , can you explain to me why the last 2 PRs I worked on (this on and #161 ) , which were both merged to 1.2 (thank you, using master in production now), the commits are not attributed to me ? 

I've worked hard for this project, partially for my personal benefit, and not being noted as a contributor, is demoralizing

**Comment by shishirmk** [comment]
@jshow taking a look.

**Comment by shishirmk** [comment]
@jshow It looks like it happened because of a squash merge. I didn't realize that it wouldn't automatically add all the authors in a squash merge.  I have added you as a co-author by amending the comment. https://github.com/Netflix/fast_jsonapi/commit/8611ea6a230eba661c8cd72514981f1ebd7a1c94

Planning to add others that have contributed to 1.2 as well. Thank you for bringing this to my attention

**Comment by jshow** [comment]
thanks @shishirmk , cheers 🍻

**Comment by gadimbaylisahil** [comment]
Hi guys. Thanks for the feature. However, adding options[:include] = [:user, :'user.first_name'] Still renders the full json for the user object instead of just first name as attribute. Is it desired behaviour? I think something wrong here

**Comment by thommahoney** [comment]
@gadimbaylisahil have a look at https://github.com/Netflix/fast_jsonapi/pull/63 or https://github.com/Netflix/fast_jsonapi/pull/254 for that functionality

**Pull request merged**